### PR TITLE
TST: signal: convert to `xp_assert_*` infrastructure, pt 2

### PR DESCRIPTION
**Pull request by ev-br**
A follow-up to https://github.com/scipy/scipy/pull/21512

- [x] test_upfirdn.py
- [x] test_cont2discrete.py
- [x] test_peak_finding.py
- [x] test_savitzky_golay.py
- [x] test_short_time_fft.py
- [x]  test_waveforms.py

### CODE REVIEW

**Comment by ev-br** [comment]
Remaining test modules: 
- `test_wavelets.py` : wavelets are deprecated, not worth bothering IMO
- `test_windows.py`  : are being worked on in https://github.com/scipy/scipy/pull/20668/
- `test_spectral.py` : are being worked on https://github.com/scipy/scipy/issues/20678

File: scipy/signal/tests/test_peak_finding.py [hunk 1]
```diff
@@ -90,15 +85,15 @@ def test_empty(self):
         """Test with empty signal."""
         x = np.array([], dtype=np.float64)
         for array in _local_maxima_1d(x):
-            assert_equal(array, np.array([]))
-            assert_(array.base is None)
+            xp_assert_equal(array, np.array([], dtype=array.dtype))
```
Comment (lucascolley) [inline scipy/signal/tests/test_peak_finding.py:88]: note to self (and below)

**Review by lucascolley — APPROVED** [review]
LGTM, thanks Evgeni!

**Review by lucascolley — CHANGES_REQUESTED** [review]
... modulo 32-bit windows

```
FAILED scipy\signal\tests\test_peak_finding.py::TestLocalMaxima1d::test_simple - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestLocalMaxima1d::test_flat_maxima - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestRidgeLines::test_single_pass - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestArgrel::test_empty - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestArgrel::test_basic - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestPeakProminences::test_basic - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestPeakProminences::test_edge_cases - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestPeakProminences::test_non_contiguous - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaks::test_plateau_size - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaks::test_height_condition - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaks::test_threshold_condition - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaks::test_distance_condition - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaks::test_prominence_condition - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaksCwt::test_find_peaks_exact - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
FAILED scipy\signal\tests\test_peak_finding.py::TestFindPeaksCwt::test_find_peaks_with_non_default_wavelets - AssertionError: dtypes do not match.
Actual: int64
Desired: int32
```

File: scipy/signal/tests/test_peak_finding.py (hunk 1 above)
Comment (ev-br) [reply scipy/signal/tests/test_peak_finding.py:88]: So basically the issue here is the recommended way to test for an empty array. What would your recommendation be?

**Comment by ev-br** [comment]
The last commit tries the following patch for 32/64-bit ints in `test_peak_finding.py::TestLocalMaxima1d::test_simple`:

```diff
diff --git a/scipy/signal/tests/test_peak_finding.py b/scipy/signal/tests/test_peak_finding.py
index 3aa50cabbb..d5c3abe50b 100644
--- a/scipy/signal/tests/test_peak_finding.py
+++ b/scipy/signal/tests/test_peak_finding.py
@@ -4,6 +4,7 @@ import numpy as np
 import pytest
 from pytest import raises, warns
 from scipy._lib._array_api import xp_assert_close, xp_assert_equal
+from scipy._lib._util import np_long
 
 from scipy.signal._peak_finding import (
     argrelmax,
@@ -99,7 +100,7 @@ class TestLocalMaxima1d:
         """Test with simple signal."""
         x = np.linspace(-10, 10, 50)
         x[2::3] += 1
-        expected = np.arange(2, 50, 3)
+        expected = np.arange(2, 50, 3, dtype=np_long)
         for array in _local_maxima_1d(x):
             # For plateaus of size 1, the edges are identical with the
             # midpoints
```

If it works on CI, will extend to other tests.

File: scipy/signal/tests/test_peak_finding.py (hunk 1 above)
Comment (lucascolley) [reply scipy/signal/tests/test_peak_finding.py:89]: I think we should just use `dtype=np.float64` here, assuming the function preserves dtype? `dtype=array.dtype` doesn't guard against regressions in the returned dtype at all.

Comment (ev-br) [reply scipy/signal/tests/test_peak_finding.py:89]: > assuming the function preserves dtype?

So in effect the suggestion is to tighten tests for the whole module. Where the dtypes have never been tested before. Wow.

**Comment by ev-br** [comment]
The remaining windows CI failure is an unrelated

```
__________ TestNumericalInverseHermite.test_basic_truncnorm_gh17155 ___________
[gw1] win32 -- Python 3.12.6 C:\hostedtoolcache\windows\Python\3.12.6\x64\python.exe
Test passed but took too long to run: Duration 1.0264428999998927s > 1.0s
```

The rest is green.

**Review by lucascolley — CHANGES_REQUESTED** [review]
thanks again!

**Review by lucascolley — APPROVED** [review]

**Pull request merged**